### PR TITLE
feat: [WD-13807] added Breadcrumb Component

### PR DESCRIPTION
**Pull request by Kxiru**
feat: [WD-13807] added Breadcrumb Component

image

[WD-13807]: https://warthogs.atlassian.net/browse/WD-13807?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: ui/src/sass/styles.scss [hunk 1]
```diff
@@ -57,3 +57,42 @@ body {
...
+  .p-breadcrumbs__items {
+    margin-bottom: 0;
+
+    .continuous-breadcrumb {
+      display: inline;
+    }
+  }
+
+  .p-breadcrumbs__item {
+    // font-size: #{map.get($font-sizes, h4-mobile)}rem;
```
Comment (mas-who) [inline ui/src/sass/styles.scss]: why is this removed from the original lxd-ui styling?

File: ui/src/sass/styles.scss [hunk 2]
```diff
@@ -57,3 +57,42 @@ body {
 .p-notification__message {
   word-break: break-word;
 }
+
+.p-breadcrumbs--large {
```
Comment (mas-who) [inline ui/src/sass/styles.scss]: let's move bread crumb styling into its own file `_breadcrumbs.scss`.

File: ui/src/components/BreadcrumbHeader.tsx [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  parentItems: ReactNode[];
+}
+
+const BreadCrumbHeader: FC<Props> = ({
+  name,
+  titleClassName,
+  parentItems,
+}: Props) => {
+  return (
+    <div className={classNames(titleClassName)}>
```
Comment (mas-who) [inline ui/src/components/BreadcrumbHeader.tsx]: Be careful with application layout here. This `div` should have the class `p-panel__title` and there should be a parent `div` with the class `p-panel__header`. Not sure if you have these classes already where you are using this component, but I think it's better to keep then contained here.

File: ui/src/components/BreadcrumbHeader.tsx [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+      <nav
+        key="breadcrumbs"
```
Comment (mas-who) [inline ui/src/components/BreadcrumbHeader.tsx]: I don't think this key is needed. We only need to add keys to components in a list when doing a mapped render.

File: ui/src/components/BreadcrumbHeader.tsx [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+            <li
+              className="p-heading--4 u-no-margin--bottom continuous-breadcrumb"
+              key={key}
+            >
+              {item}
+            </li>
+          ))}
+
+          <li
+            className="p-heading--4 u-no-margin--bottom name continuous-breadcrumb"
```
Comment (mas-who) [inline ui/src/components/BreadcrumbHeader.tsx]: the `name` class here comes from lxd-ui  in the `_rename_header.scss` file. I think it's important to apply the same in this project. I think it can all go into the breadcrumbs scss file. Also we should have a tooltip here similar to lxd-ui since the name may be very long and would be clipped if it overflows (see styling in lxd-ui).

**Review by mas-who — COMMENTED** [review]
nice one splitting this out into a separate PR. It's much easier to focus on it. Left some feedback

File: ui/src/sass/styles.scss (hunk 1 above)
Comment (Kxiru) [reply ui/src/sass/styles.scss]: it returned an error and I could not identify where it was imported from.

File: ui/src/components/BreadcrumbHeader.tsx (hunk 5 above)
Comment (Kxiru) [reply ui/src/components/BreadcrumbHeader.tsx]: Should we leave the rename header styling and functionality out for now, until we are ready to implement it fully?

Happy to add the tooltip! (and remove name)

File: ui/src/components/BreadcrumbHeader.tsx (hunk 3 above)
Comment (Kxiru) [reply ui/src/components/BreadcrumbHeader.tsx]: Since they're passed through as the prop to BaseLayout and therefore AppMain => Panel => It should inherit that styling.

File: ui/src/sass/styles.scss (hunk 1 above)
Comment (mas-who) [reply ui/src/sass/styles.scss]: I think it's the sass map function that we are not importing causing the issue. Try put `@use "sass:map";` at the top of the file where you use `map.get`.

File: ui/src/components/BreadcrumbHeader.tsx (hunk 5 above)
Comment (mas-who) [reply ui/src/components/BreadcrumbHeader.tsx]: I think we don't need to move all the styling over. The most important thing is to check what happens when the name is very long. At least we need to cater for text overflow, since for the current use case, cluster name is set by a user so we don't know how long it can get.

File: ui/src/components/BreadcrumbHeader.tsx (hunk 3 above)
Comment (mas-who) [reply ui/src/components/BreadcrumbHeader.tsx]: right, in that case I don't think you need this `div`. You can use the `titleComponent`, `title`, and `titleClassName` props on the `Panel` component from react-components to achieve the same layout as lxd-ui. You'd have to adjust `BaseLayout` a bit though since currently it does not allow input of any other panel props.

File: ui/src/components/BreadcrumbHeader.tsx [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+              className="p-heading--4 u-no-margin--bottom continuous-breadcrumb"
+              key={key}
+            >
+              {item}
+            </li>
+          ))}
+
+          <li
+            className="p-heading--4 u-no-margin--bottom continuous-breadcrumb"
+            title={`${name}`}
```
Comment (mas-who) [inline ui/src/components/BreadcrumbHeader.tsx]:
```suggestion
            title={name}
```

File: ui/src/components/BreadcrumbHeader.tsx (hunk 5 above)
Comment (Kxiru) [reply ui/src/components/BreadcrumbHeader.tsx]: I see what you mean. Please let me know what you think of this next review.

File: ui/src/components/BreadcrumbHeader.tsx [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        {parentItems ? (
+          parentItems.map((item, key) => (
+            <li
+              className="p-heading--4 u-no-margin--bottom continuous-breadcrumb"
+              key={key}
+            >
+              {item}
+            </li>
+          ))
+        ) : (
```
Comment (mas-who) [inline ui/src/components/BreadcrumbHeader.tsx]: better to set false condition for ternary as `null`

File: ui/src/components/BaseLayout.tsx [hunk 8]
```diff
@@ -23,6 +23,7 @@ const BaseLayout: FC<Props> = ({
         controls={controls}
         wrapContent={true}
         contentClassName={contentClassName}
+        titleClassName="base-layout-header"
```
Comment (mas-who) [inline ui/src/components/BaseLayout.tsx]: I don't see this class defined anywhere? Also I think `titleClassName` should be passed in as a prop in case if we need to customize styling for the related `title` prop.

The other prop from the `Panel` component we need here is [titleComponent](https://github.com/canonical/react-components/blob/main/src/components/Panel/Panel.tsx). It is by default set to `h4` in `Panel`, but if we want to recreate what's in lxd-ui, it needs to be a `div`.

File: ui/src/pages/clusters/ClusterList.tsx [hunk 9]
```diff
@@ -20,7 +21,11 @@ const ClusterList: FC = () => {
   const tabs: string[] = ["Active", "Pending", "Tokens"];
 
   return (
-    <BaseLayout title={"Clusters"} controls={<AddClusterButton />}>
+    <BaseLayout
```
Comment (mas-who) [inline ui/src/pages/clusters/ClusterList.tsx]: we shouldn't do this change. For a top level page component like `ClusterList`, setting `title` to `BreadCrumbHeader` means we are introducing unnecessary html. Same thing goes for the change in `Settings`

**Review by mas-who — COMMENTED** [review]
almost there, thanks for bring the `.name` styles over :)

**Comment by Kxiru** [comment]
Thanks, @mas-who , let me know your thoughts now :)

**Comment by mas-who** [comment]
LGTM thanks for the changes :+1:

**Review by mas-who — APPROVED** [review]
